modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDBKeyValueStorage.java

```diff
@@ -0,0 +1,1093 @@
+    private long rev;
+
+    /** Update counter. Will be incremented for each update of any particular entry. */
+    private long updCntr;
+
+    /**
+     * Constructor.
+     *
+     * @param dbPath RocksDB path.
+     */
+    public RocksDBKeyValueStorage(Path dbPath) {
+        try {
+            options = new DBOptions()
+                .setCreateMissingColumnFamilies(true)
+                .setCreateIfMissing(true);
+
+            this.dbPath = dbPath;
+
+            Options dataOptions = new Options().setCreateIfMissing(true)
+                // The prefix is the revision of an entry, so prefix length is the size of a long
+                .useFixedLengthPrefixExtractor(Long.BYTES);
+
+            ColumnFamilyOptions dataFamilyOptions = new ColumnFamilyOptions(dataOptions);
+
+            Options indexOptions = new Options().setCreateIfMissing(true);
+
+            ColumnFamilyOptions indexFamilyOptions = new ColumnFamilyOptions(indexOptions);
+
+            List<ColumnFamilyDescriptor> descriptors = Arrays.asList(
+                new ColumnFamilyDescriptor(DATA.nameAsBytes(), dataFamilyOptions),
+                new ColumnFamilyDescriptor(INDEX.nameAsBytes(), indexFamilyOptions)
+            );
+
+            var handles = new ArrayList<ColumnFamilyHandle>();
+
+            // Delete existing data, relying on the raft's snapshot and log playback
+            destroyRocksDB();
+
+            this.db = RocksDB.open(options, dbPath.toAbsolutePath().toString(), descriptors, handles);
+
+            data = new ColumnFamily(db, handles.get(0), DATA, dataFamilyOptions, dataOptions);
+
+            index = new ColumnFamily(db, handles.get(1), INDEX, indexFamilyOptions, indexOptions);
+        }
+        catch (Exception e) {
+            try {
+                close();
+            }
+            catch (Exception exception) {
+                e.addSuppressed(exception);
+            }
+
+            throw new IgniteInternalException("Failed to start the storage", e);
+        }
+    }
+
+    /**
+     * Clear the RocksDB instance.
+     * The major difference with directly deleting the DB directory manually is that
+     * destroyDB() will take care of the case where the RocksDB database is stored
+     * in multiple directories. For instance, a single DB can be configured to store
+     * its data in multiple directories by specifying different paths to
+     * DBOptions::db_paths, DBOptions::db_log_dir, and DBOptions::wal_dir.
+     *
+     * @throws RocksDBException If failed.
+     */
+    private void destroyRocksDB() throws RocksDBException {
+        try (final Options opt = new Options()) {
+            RocksDB.destroyDB(dbPath.toString(), opt);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() throws Exception {
+        IgniteUtils.closeAll(options, data, index, db);
+    }
+
+    /** {@inheritDoc} */
+    @NotNull
+    @Override public CompletableFuture<Void> snapshot(Path snapshotPath) {
```

Comment:
what will happen if multiple snapshots will be created at the same time? Or is this situation impossible?

Reply:
Nothing bad at all, we'll just use the snapshot of the database at the time when a snapshot was requested. 
I'm not sure though that this situation won't happen, but afaik it should not.

Reply:
I'm more concerned that they will start writing to the same tmp file. Or am I missing something?

Reply:
Raft takes care of the directories by creating a new one for every snapshot